dipp/tensorflow/unet.py

```diff
@@ -0,0 +1,251 @@
+
+            if use_batch_norm and not disable_batch_norm:
+                out = tf.contrib.layers.batch_norm(out,
+                                                   is_training=is_training)
+            if keep_prob != 1.0 and not disable_dropout:
+                out = tf.contrib.layers.dropout(out, keep_prob=keep_prob,
+                                                is_training=is_training)
+
+            if not disable_activation:
+                out = apply_activation(out)
+
+            return out
+
+    def apply_maxpool(x):
+        if ndim == 1:
+            return maxpool1d(x)
+        else:
+            return maxpool2d(x)
+
+    finals = []
+
+    with tf.name_scope('{}_call'.format(name)):
+        with tf.name_scope('in'):
+            current = apply_conv(x, features)
+            current = apply_conv(current, features)
+            finals.append(current)
+
+        for layer in range(depth - 1):
+            with tf.name_scope('down_{}'.format(layer + 1)):
+                features_layer = 2 ** (layer + 1)
+                current = apply_maxpool(current)
+                current = apply_conv(current, features_layer)
+                current = apply_conv(current, features_layer)
+                finals.append(current)
+
+        with tf.name_scope('coarse'):
+            current = apply_maxpool(current)
+            current = apply_conv(current, features * 2 ** depth)
+            current = apply_conv(current, features * 2 ** depth)
+
+        for layer in reversed(range(depth - 1)):
+            with tf.name_scope('up_{}'.format(layer + 1)):
+                features_layer = 2 ** (layer + 1)
+                skip = finals.pop()
+                current = apply_convtransp(current, features_layer,
+                                           out_shape=tf.shape(skip),
+                                           disable_activation=True)
+                current = tf.concat([current, skip], axis=-1)
+
+                current = apply_conv(current, features_layer)
+                current = apply_conv(current, features_layer)
+
+        with tf.name_scope('out'):
+            skip = finals.pop()
+            current = apply_convtransp(current, features,
+                                       out_shape=tf.shape(skip),
+                                       disable_activation=True)
+            current = tf.concat([current, skip], axis=-1)
+
+            current = apply_conv(current, features)
+            current = apply_conv(current, features)
+
+            current = apply_conv(current, nout,
+                                 size=1,
+                                 disable_activation=True,
+                                 disable_batch_norm=True,
+                                 disable_dropout=True)
+
+    return current
+
+
+if __name__ == '__main__':
+    from dipp.util.testutils import run_doctests
+    with tf.Session():
+        run_doctests()
```

Comment:
You may want to add `extraglobs={'tf': tf}` here as well as to the `add_doctest_modules` in `pytest_plugins.py`.